src/daft-core/src/series/ops/arithmetic.rs

```diff
@@ -22,11 +23,262 @@ macro_rules! impl_arithmetic_for_series {
     };
 }
 
-impl_arithmetic_for_series!(Add, add);
-impl_arithmetic_for_series!(Sub, sub);
-impl_arithmetic_for_series!(Mul, mul);
-impl_arithmetic_for_series!(Div, div);
-impl_arithmetic_for_series!(Rem, rem);
+macro_rules! binary_op_unimplemented {
+    ($lhs:expr, $op:expr, $rhs:expr, $output_ty:expr) => {
+        unimplemented!(
+            "No implementation for {} {} {} -> {}",
+            $lhs.data_type(),
+            $op,
+            $rhs.data_type(),
+            $output_ty,
+        )
+    };
+}
+
+impl Add for &Series {
+    type Output = DaftResult<Series>;
+    fn add(self, rhs: Self) -> Self::Output {
+        let output_type =
+            InferDataType::from(self.data_type()).add(InferDataType::from(rhs.data_type()))?;
+        let lhs = self;
+        match &output_type {
+            #[cfg(feature = "python")]
+            DataType::Python => run_python_binary_operator_fn(lhs, rhs, "add"),
+            DataType::Utf8 => {
+                Ok(cast_downcast_op!(lhs, rhs, &DataType::Utf8, Utf8Array, add)?.into_series())
```

Comment:
so this was likely already an issue, but won't this cause us to do a bunch of extra work casting and downcasting when datatypes match the output type?

Reply:
The cast logic should check if it already matches the target dtype